libs/langchain/langchain/chat_models/sparkllm.py

```diff
@@ -0,0 +1,479 @@
+        except ImportError:
+            raise ImportError(
+                "Could not import websocket client python package. "
+                "Please install it with `pip install websocket-client`."
+            )
+
+        self.api_url = (
+            "wss://spark-api.xf-yun.com/v3.1/chat" if not api_url else api_url
+        )
+        self.app_id = app_id
+        self.ws_url = _SparkLLMClient._create_url(
+            self.api_url,
+            api_key,
+            api_secret,
+        )
+        self.model_kwargs = model_kwargs
+        self.spark_domain = spark_domain or "generalv3"
+        self.queue: Queue[Dict] = Queue()
+        self.blocking_message = {"content": "", "role": "assistant"}
+
+    @staticmethod
+    def _create_url(api_url: str, api_key: str, api_secret: str) -> str:
+        """
+        Generate a request url with an api key and an api secret.
+        """
+        # generate timestamp by RFC1123
+        date = format_date_time(mktime(datetime.now().timetuple()))
+
+        # urlparse
+        parsed_url = urlparse(api_url)
+        host = parsed_url.netloc
+        path = parsed_url.path
+
+        signature_origin = f"host: {host}\ndate: {date}\nGET {path} HTTP/1.1"
+
+        # encrypt using hmac-sha256
+        signature_sha = hmac.new(
+            api_secret.encode("utf-8"),
+            signature_origin.encode("utf-8"),
+            digestmod=hashlib.sha256,
+        ).digest()
+
+        signature_sha_base64 = base64.b64encode(signature_sha).decode(encoding="utf-8")
+
+        authorization_origin = f'api_key="{api_key}", algorithm="hmac-sha256", \
+        headers="host date request-line", signature="{signature_sha_base64}"'
+        authorization = base64.b64encode(authorization_origin.encode("utf-8")).decode(
+            encoding="utf-8"
+        )
+
+        # generate url
+        params_dict = {"authorization": authorization, "date": date, "host": host}
+        encoded_params = urlencode(params_dict)
+        url = urlunparse(
+            (
+                parsed_url.scheme,
+                parsed_url.netloc,
+                parsed_url.path,
+                parsed_url.params,
+                encoded_params,
+                parsed_url.fragment,
+            )
+        )
+        return url
+
+    def run(
+        self,
+        messages: List[Dict],
+        user_id: str,
+        model_kwargs: Optional[dict] = None,
+        streaming: bool = False,
+    ) -> None:
+        self.websocket_client.enableTrace(False)
+        ws = self.websocket_client.WebSocketApp(
+            self.ws_url,
+            on_message=self.on_message,
+            on_error=self.on_error,
+            on_close=self.on_close,
+            on_open=self.on_open,
+        )
+        ws.messages = messages
+        ws.user_id = user_id
+        ws.model_kwargs = self.model_kwargs if model_kwargs is None else model_kwargs
+        ws.streaming = streaming
+        ws.run_forever(sslopt={"cert_reqs": ssl.CERT_NONE})
```

Comment:
Why is `ssl.CERT_NONE` required here? Are there certificate issues on the domain?

Reply:
I had certificate issue before with their api, but haven't had this problem recently. I'll remove the parameter and test it, then submit my fix.